pyiceberg/table/__init__.py

```diff
@@ -350,6 +357,241 @@ class RemovePropertiesUpdate(TableUpdate):
     removals: List[str]
 
 
+class TableMetadataUpdateContext:
+    updates: List[TableUpdate]
+    last_added_schema_id: Optional[int]
```

Comment:
It looks like we don't use this one?
```suggestion
```

Reply:
There is one case that we need this:
For `SetCurrentSchemaUpdate`, if the given `schema_id` is `-1`, we will set the current schema to last added schema by the stored `last_added_schema_id`.

Reply:
I prefer to keep them separate. I think we might need to have some additional checks here such as, what happens if you add a column, and then revoke the column again. It will first create a new schema, with a new ID, and then it will reuse the old schema again.

- 1: Schema(a: int), current_schema_id=1
Add column b:
- 1: Schema(a: int), 2: Schema(a: int, b: int), current_schema_id=2
Drop column b:
- 1: Schema(a: int), 2: Schema(a: int, b: int), current_schema_id=1

Reply:
Thanks for sharing the example! I'd like to clarify some points to ensure I fully understand your perspective. According to [Java ref](https://github.com/apache/iceberg/blob/2aac63688054ac36540072cfa13aac08fccf9421/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1449-L1489), It appears that `lastAddedSchemaId` refers to the most recently added schema in the **current** transaction.

The purpose of `lastAddedSchemaId `in Java is to support multiple `AddSchema` updates within a single transaction, allowing for efficient tracking of the latest schema addition. This is particularly beneficial if a schema added later in the transaction is the same as one added earlier. However, in Python, where we limit to one `AddSchema` per transaction, `lastAddedSchemaId` should either be the ID of the added schema or None, specifically when the added schema is identical to existing ones in the metadata.

The example you provided brings up an important consideration for our Python implementation: if the result schema of `UpdateSchema` is identical to a previously added schema in base_metadata, we should not add this new schema to the metadata. Instead, we only need to set the current schema ID to that of the previously existing one.
In Java, metadata builder will suppress the `AddSchema` update if the new schema is a duplicate of any previous schemas. Inspired by this, I think we can implement a similar mechanism in `UpdateSchema.commit()`. If the new schema is identical to an existing one, only a single `setCurrentSchema` should be added.

In conclusion, to enhance our approach to schema updates, I propose the following:
1. In `UpdateSchema.commit()`: We should check if new_schema is identical to any existing schemas before adding an `AddSchema` and a `SetCurrentSchema`. If they are identical, only a `SetCurrentSchema(existing_schema_id)` is necessary.
2. Regarding `lastAddedSchemaId`: It’s crucial to ensure that if it's not None, it points to a schema added earlier in the current transaction. Since Python transactions contain only one AddSchema, this should be straightforward. We could also consider adding a uniqueness check in update_table_metadata to confirm that the updates are from a single transaction. Also, instead of having a `lastAddedSchemaId`, we may add a method called `last_added_schema()` to extract the last added schema from the update context.

What do you think – do the above reasoning and proposed changes seem good to you?

Thanks!

Reply:
Thanks for raising this, this is currently not handled correctly on the rest side either.

I've created a PR [here](https://github.com/apache/iceberg-python/pull/176) as a suggestion, but also feel free to fix it in your PR.

Reply:
@HonahX can you rebase? I think this one is ready to go!

Reply:
@Fokko Thanks for reviewing! I've pushed my updates (mostly on `SetCurrentSchemaUpdate`) and rebased the PR.